src/main/java/com/fasterxml/jackson/databind/deser/std/StdKeyDeserializer.java

```diff
@@ -410,9 +410,11 @@ private EnumResolver _getToStringResolver(DeserializationContext ctxt)
             EnumResolver res = _byToStringResolver;
             if (res == null) {
                 synchronized (this) {
-                    res = EnumResolver.constructUsingToString(ctxt.getConfig(),
+                    if (res == null) {
```

Comment:
actually, I think the `res` local variable would be better removed and references to `_byToStringResolver ` used instead

Reply:
Have we considered using AtomicReferences instead? A good example that could be used as a template here is the first answer in https://stackoverflow.com/questions/20087173/how-to-do-a-lazy-create-and-set-with-atomicreference-in-a-safe-and-efficient-man

Reply:
sure - ignore comment about AtomicReference and just use the volatile references with code synchronization

Reply:
@pjfanning i think atomicref is worse here than the standard volatile double-checked locking from shipilevs article:

```java
public class SafeLocalDCLFactory implements Factory {
  private volatile Singleton instance;

  @Override
  public Singleton getInstance() {
    Singleton res = instance;
    if (res == null) {
      synchronized (this) {
        res = instance;
        if (res == null) {
          res = new Singleton();
          instance = res;
        }
      }
    }
    return res;
  }
}
```

one volatile read on the happy path, that's the least overhead you can get without relying on final field semantics.

Reply:
@JooHyukKim can you use this latest approach?

This type of code is complicated. Scala has `lazy val` but it can deadlock and they've been trying to fix it in recent Scala3 versions and the code that gets generated under the hood keeps getting more complicated.

Reply:
Latest approach, you mean using AtomicReference? @pjfanning

Reply:
latest - 1 comment above my comment - comment from @yawkat about `public Singleton getInstance()`

Reply:
@pjfanning I see, looks cleaner

Reply:
I was thinking create one SingleFactory class to host `getInstance()` for both `_byToStringResolver` and `_getToStringLookup`. 

How do you think?

Reply:
It's just this bit we want you use

```
Singleton res = instance;
    if (res == null) {
      synchronized (this) {
        res = instance;
        if (res == null) {
          res = new Singleton();
          instance = res;
        }
      }
    }
    return res;
```

Reply:
I see I see, on it